[RegAllocEvictAdvisor] Add minimum weight ratio heuristic.

mgudim opened the pull request:
Do not evict a live range if eviction will break existing hint without satisfying a new one and the ratio of weights is not big enough.

mgudim committed:
[RegAllocEvictAdvisor][NFC] Add minimum weight ratio heuristic.

Do not evict a live range if eviction will break existing hint without
satisfying a new one and the ratio of weights is not big enough. The
heuristic is disabled by default.

mgudim committed:
[RegAllocEvictAdvisor][NFC] Add minimum weight ratio heuristic.

Do not evict a live range if eviction will break existing hint without
satisfying a new one and the ratio of weights is not big enough. The
heuristic is disabled by default.

mtrofin commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -44,6 +44,12 @@ static cl::opt<bool> EnableLocalReassignment(
+    "min-weight-ratio-needed-to-evict-hint", cl::Hidden,
+    cl::desc("The minimum ration of weights needed in order for the live range with bigger weight to evict the other live range which"
+    "satisfies a hint),
+    cl::init(7.5));
```
This should default to the value that makes this change a NFC, so `-inf`?

qcolombet replied:
To add on what @mtrofin said, could you explain the rationale for `7.5`?

It feels random to me.

mtrofin commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -157,6 +163,10 @@ bool DefaultEvictionAdvisor::shouldEvict(const LiveInterval &A, bool IsHint,
     return true;
 
   if (A.weight() > B.weight()) {
+    float WeightRatio = A.weight() / B.weight();
```
can `B.weight()` be 0?

mgudim replied:
fixed

mtrofin reviewed (commented):
Shouldn't there be a test?

mgudim commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -44,6 +44,12 @@ static cl::opt<bool> EnableLocalReassignment(
 
+static cl::opt<float> MinWeightRatioNeededToEvictHint(
+    "min-weight-ratio-needed-to-evict-hint", cl::Hidden,
+    cl::desc("The minimum ration of weights needed in order for the live range with bigger weight to evict the other live range which"
```
Fixed, thanks!

mgudim commented:
> Shouldn't there be a test?

Added test

mgudim commented:
@mtrofin @qcolombet 

This patch makes a lot of difference only on top of this: https://github.com/llvm/llvm-project/pull/90819
It actually gives good improvements in terms of dynamic instruction count on gcc and perlbench benchmarks.
I choose the number `7.5` by experimenting on xz where I originally had a degradation, but with 7.5 it's actually a slight improvement. Turned out the same `7.5` worked pretty well for the rest of the spec benchmarks. Perhaps a better number exists, but this needs a lot of experimentation. Maybe we should make the number architecture-dependent or even set it programatically depending on the code compiled? For now the value of `1.0` disables the heuristic by default.

mgudim commented:
I've looked at the CI error - it is some assertion triggered from SLP vectorizer while compiling something in libcxx. I can't imagine how this change could trigger this, so I suspect something in the upstream is broken.

mgudim commented:
@mtrofin @qcolombet After rebasing, the CI passes. Is this good to merge?

qcolombet commented on llvm/test/CodeGen/RISCV/min-weight-ratio.mir:
```diff
@@ -0,0 +1,76 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc %s -mtriple=riscv64 -run-pass=greedy,virtregrewriter \
+# RUN: -min-weight-ratio-needed-to-evict-hint=7.5 -o - | FileCheck %s
```
To go back to what I was saying wrt 7.5. I'm not questioning that this gives you the best outcome in what you measured, what I was asking is why it is.
Essentially what is the explanation that makes 7.5 a good number aside from "it fixes my problem".

mgudim committed:
[CodeGen] Improve heuristic for eviction decisions.

Currently, the `DefaultEvictionAdvisor::shouldEvict` makes the decision
to evict based on `weight`. The higher the use-def frequency of a live
range, the harder it should be to evict a live range, since it is more
"complex". Conversely, the bigger the size of a live range, the easier
it should be to evict, since the eviction would free up a register for a
longer duration. The `weight` combines the two numbers into one by
dividing use-def frequency by the size.

Turns our that a better way to combine the two exists. When one live range
has a significantly higher use-def frequence, we should just ignore the
sizes. If use-def frequencies are comparable, we should evict only if
the size of evictee is much larger that the size of evictor.

mgudim committed:
set default `MinSizeRatioNeededToEvict` to 0.05

mgudim committed:
format, remove test

mgudim committed:
Only enable the new heuristic if the flag is passed.

This avoids modifying many codegen tests.

qcolombet commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```suggestion
    cl::desc("Make eviction decisions based on simple weight comparison"),
```

qcolombet replied:
Don't know if it will become clearer later in the PR yet, but we should explain somewhere what "simple weight" means.

mgudim replied:
Here "simple" is attached to "comparison", not "weight". I wanted to say "we simply compare the weights of A and B" as opposed to the new heuristic

qcolombet commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
What does "use-def frequency" mean?
It this the cumulated frequencies of all of A's uses and defs?

mgudim replied:
yes

qcolombet commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
The range seems a bit random to me.
Can you explain the rationale behind `1/MinUseDefFreqRatioToIgnoreSize`?
Like for the upper bound, what you're saying is if A is used/defined more than 5x than B, we ignore how much the live-range spans across the program (aka its size).

mgudim replied:
yes, exactly. I.e. the same upper bound definition is symmetric wrt A and B.

qcolombet commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```suggestion
        "ignore the sizes of A and B when making the eviction decision."),
```

qcolombet reviewed (commented):
My general comment didn't change.
I'm not questioning that you're getting better results in your benchmark, but I still would like to understand why this is the case.
These numbers seem still too magic to me.
